api/peering_test.go

```diff
@@ -2,14 +2,155 @@ package api
 	"testing"
 	"time"
 
+	"github.com/hashicorp/serf/testutil/retry"
 	"github.com/stretchr/testify/require"
 
 	"github.com/hashicorp/consul/sdk/testutil"
 )
 
+func peerExistsInPeerListings(peer *Peering, peerings []*Peering) bool {
+	for _, aPeer := range peerings {
+		isEqual := (peer.PeerID == aPeer.PeerID) &&
+			(reflect.DeepEqual(peer.PeerCAPems, aPeer.PeerCAPems)) &&
+			(peer.PeerServerName == aPeer.PeerServerName) &&
+			(peer.Partition == aPeer.Partition) &&
+			(peer.Name == aPeer.Name) &&
+			(reflect.DeepEqual(peer.PeerServerAddresses, aPeer.PeerServerAddresses)) &&
+			(peer.State == aPeer.State) &&
+			(peer.CreateIndex == aPeer.CreateIndex) &&
+			(peer.ModifyIndex) == aPeer.ModifyIndex
+
+		if isEqual {
+			return true
+		}
+	}
+
+	return false
+}
+
+func TestAPI_Peering_Read(t *testing.T) {
+	t.Parallel()
+	c, s := makeClientWithCA(t)
+	defer s.Stop()
+	s.WaitForSerfCheck(t)
+	ctx := context.Background()
+	peerings := c.Peerings()
+
+	t.Run("call Read with no name", func(t *testing.T) {
+		resp, qm, err := peerings.Read(ctx, PeeringRequest{}, nil)
+
+		// basic checks
+		require.EqualError(t, err, "peering name cannot be empty")
+		require.Empty(t, qm)
+		require.Empty(t, resp)
+	})
+
+	t.Run("read token that does not exist on server", func(t *testing.T) {
+		resp, qm, err := peerings.Read(ctx, PeeringRequest{Name: "peer1"}, nil)
+
+		// basic checks
+		require.NotNil(t, err) // 404
+		require.Empty(t, qm)
+		require.Empty(t, resp)
+	})
+
+}
+
+// TestAPI_Peering_List
+func TestAPI_Peering_List(t *testing.T) {
+	t.Parallel()
+	c, s := makeClientWithCA(t)
+	defer s.Stop()
+	s.WaitForSerfCheck(t)
+	ctx := context.Background()
+	peerings := c.Peerings()
+
+	t.Run("call List when no peers should exist", func(t *testing.T) {
+		resp, qm, err := peerings.List(ctx, PeeringListRequest{}, nil)
+
+		// basic checks
+		require.NoError(t, err)
+		require.NotEmpty(t, qm)
+
+		require.Empty(t, resp) // no peerings so this should be empty
+	})
+
+	t.Run("call List when peers are present", func(t *testing.T) {
+		// Generate a token happy path
+		resp, wm, err := peerings.GenerateToken(ctx, PeeringGenerateTokenRequest{PeerName: "peer1"}, nil)
+
+		require.NoError(t, err)
+		require.NotEmpty(t, wm)
+		require.NotEmpty(t, resp)
+
+		peering, qm, err2 := peerings.Read(ctx, PeeringRequest{Name: "peer1"}, nil)
+
+		// basic ok checking
+		require.NoError(t, err2)
+		require.NotEmpty(t, qm)
+		require.NotEmpty(t, peering)
+
+		peeringsList, qm, err := peerings.List(ctx, PeeringListRequest{}, nil)
+
+		// basic checks
+		require.NoError(t, err)
+		require.NotEmpty(t, qm)
+
+		require.True(t, peerExistsInPeerListings(peering, peeringsList), "expected to find peering in list response")
+
+		// modify peering to some non existent peering
+		peering.Name = "not_peer1"
+
+		require.False(t, peerExistsInPeerListings(peering, peeringsList), "did not expect to find peering in list response")
+	})
+
+	t.Run("call List with dc1", func(t *testing.T) {
```

Comment:
This test can be removed; the user never actually interacts with the datacenter; it gets overwritten with the agent's dc in the handler